[[ Win64 ]] Initial changes for Win32 x86-64

runrevmark opened the pull request:
This patch contains initial changes for Win32 x86-64 builds.

Depends on: https://github.com/runrevmark/livecode/pull/10

livecodefraser committed:
[[ Win64 ]] Update engine to compile for win32 x86-64

This patch tweaks the engine code to compile for win32 x86-64.

Mainly it uses intptr_t in the gentle related toolchain code
rather than long, and DWORD_PTR / LONG_PTR appropriately in
Win32 APIs.

Additionally, it updates the build-libs.bat script to use v140
toolset (even if VS 15 is requested - this is because VS2017 does
not have the same structure, and it only bumps the toolset version
by 1, not 10 - i.e. v141 and not v150).

Finally, it modifies fetch-libraries.sh to ensure that the win32
prebuilts have the correct names. The system expects the toolset
to be v141, and for 64-bit it expects the arch to be x64.

runrevmark committed:
[[ Win64 ]] Update mergExt pointer

This patch updates the mergExt pointer in the builder_utilities
stack to point to the version containing preliminary 64-bit
builds of the Windows mergExt externals.

peter-b commented on config/win32.gypi:
```diff
@@ -25,6 +25,18 @@
+		'library_dirs':
+		[
+			'$(foo)C:/WinDDK/7600.16385.1/lib/ATL/i386',
+			'$(foo)C:/WinDDK/7600.16385.1/lib/MFC/i386',
```
Should these be conditional on Windows architecture name?

runrevmark replied:
I don't think those are necessary - so I'm trying removing them. The default setup of Vs2015/2017 seems to work fine without them.

peter-b commented on configure.bat:
```diff
@@ -22,6 +22,13 @@ IF %TARGET_ARCH%==x86_64 (
+) ELSE (
+    SET LC_ARCH=x86
+)
+
```
Why is `%LC_ARCH%` needed? I think the intention with this script was that `%TARGET_ARCH%` should be the LiveCode architecture name, given the way that it is used to compute `%MSVC_ARCH%` (please correct me if that is incorrect!)

runrevmark replied:
There are still things which rely on TARGET_ARCH being the (native) arch name. This needs a bit of a moomin - the current configure.bat ensures that MSVC_ARCH and TARGET_ARCH are both set to X64 for 64-bit, where as LC_ARCH uses X86_64.

peter-b commented on engine/src/combiners.cpp:
```diff
@@ -301,32 +301,9 @@ static INLINE uint16_t long_scaled_divide(uint32_t n, uint8_t s, uint16_t d)
-#else
 	return 1;
-#endif
 }
```
Maybe this function should be removed entirely; it no longer does anything useful on any platform.

runrevmark replied:
Possibly - I suspect this file and legacyblendmodes.cpp are going to be obsoleted by the Skia changes... So, I think dealing with it there is a better idea.

peter-b commented on libcore/src/core.cpp:
```diff
@@ -66,6 +66,7 @@ bool MCThrow(uint32_t p_error)
 #include <crtdbg.h>
 #include <dbghelp.h>
 
+// WIN64-CHECK
```
Does this comment indicate that some additional work is required in the future?  If so, it should probably contain `TODO`.

runrevmark replied:
They were just markers for things I needed to double-check - will do so and remove :)

peter-b commented on libgraphics/src/legacyblendmodes.cpp:
```diff
@@ -375,29 +375,7 @@ static INLINE uint16_t long_scaled_divide(uint32_t n, uint8_t s, uint16_t d)
-#else
 	return 1;
-#endif
 }
```
As previously, I think this function would be best removed entirely.

runrevmark replied:
See above

runrevmark committed:
[[ Win64 ]] Update tsNet pointer

This patch updates the tsNet version pointer to the first version
containing preliminary 64-bit Win32 builds.

runrevmark committed:
Merge branch 'develop' into win64

runrevmark committed:
[[ CEF ]] Update CEF to 2704 and add 64-bit Win32 support

This patch updates the CEF version to 2704, and patches the weak
stub maker to cope with an intsize type so that stub generation
works for 64-bit Win32.

runrevmark committed:
[[ Win64 ]] Remove unnecessary library_dir config

This patch removes unnecessary setting of library_dirs specific
to windows builds.

runrevmark committed:
[[ Win64 ]] Remove unnecessary extern in libcore

This patch removes an unnecessary Win32 API extern from
libcore/core.cpp

runrevmark committed:
[[ Win64 ]] Remove unnecessary atlthunk.lib from revbrowser

This patch removes a seemingly unnecessary reference to atlthunk.lib
from revbrowser.

runrevmark committed:
[[ Win64 ]] Make sure stdint.h is included in gen.h for toolchain

This patch ensures that 'gen.h' files generated by gentle include
stdint.h before defining a struct containing intptr_t.

runrevmark committed:
Merge branch 'win64' into update-cef_2704

Commit:
[[ CEF ]] Tweak targets of Linux CEF files

This patch adjusts the copy files rules for CEF resources copied
into the build products dir for running from source.

runrevmark committed:
[[ CEF ]] Update version to a more recent build of 2704

This patch updates the CEF version to use a build not using
tcmalloc on Linux.

runrevmark committed:
Merge branch 'update-cef_2704' of https://github.com/runrevmark/livecode into update-cef_2704

runrevmark committed:
[[ CEF ]] Disable logging

Turn off CEF logging to stop immense amount of output to stderr.

runrevmark committed:
[[ Win64 ]] Add win32-x86_64 triple to buildbot.py

This patch enables 64-bit win32 builds for this branch by adding
the triple to buildbot's KNOWN_TRIPLES.

runrevmark committed:
Merge branch 'win64' into update-cef_2704

runrevmark committed:
[[ CEF ]] Make sure icudat and blobs are next to main exe on Linux

This patch ensures that files which have to be next to the main
executable for CEF to run are so on Linux.

runrevmark committed:
Merge pull request #10 from runrevmark/update-cef_2704

[[ CEF ]] Update CEF to 2704 and add 64-bit Win32 support

runrevmark committed:
[[ Win64 ]] Correct comparison in MakeIntegerLiteral

This patch corrects a signed/unsigned comparison which was causing
the test for the size of integer literals allowed in LCB source
files to fail.

runrevmark committed:
[[ Win64 ]] Sort out arch vars, and ensure prebuilts are fetched

This patch ensures that TARGET_ARCH is always x86_64 for 64-bit
X86 processors, and that target_arch (the gyp variable) is always
x64 for x86-64 on Win32. Additionally, it introduces a new gyp var
uniform_arch, which is either x86_64 or x86.

Finally, a 'fetch' phase has been added which fetches the prebuilts
for host and target. This ensures that they are present before
configuration (which eliminates missing file warnings from gyp) and
that Visual Studio can build projects in its own order (which can
cause failure to find prebuilt includes).

runrevmark committed:
[[ Win64 ]] Make sure 'fetch' step is understood by buildbot.py

This patch adds 'fetch' to the list of understood steps in
buildbot.py to make sure it is invoked.

runrevmark committed:
[[ Win64 ]] Fix missing import in fetch.py

This patch ensures the RE module is imported in fetch.py.

runrevmark committed:
[[ Win64 ]] Fix toolset patching for prebuilts

This patch fixes the Win32 specific patching fetch-libraries.sh
does to provide v141 builds from v140 builds.

runrevmark committed:
[[ Win64 ]] Fix mapping of CEF prebuilt name for Win32

This patch ensures the CEF archive name is correct for x86 and
x86_64.

runrevmark committed:
Merge branch 'develop' into win64

runrevmark committed:
[[ Win64 ]] Remove wow_helper from 32-bit CEF copy files

This patch removes the now unneeded wow_helper.exe from the 32-bit
libcef copy files action.

runrevmark committed:
Merge branch 'develop' into win64

runrevmark committed:
[[ Win64 ]] Make sure test_new.cpp compiles in VC

This patch ensures that test_new.cpp compiles in VC which has a
limit of 0x7fffffff bytes for any declared array.

runrevmark committed:
[[ Windows ]] Make sure all windows set main window as owner on Win32

This patch ensures that all windows the engine owns sets their
Owner to the main window (if any) before they are opened.

runrevmark committed:
[[ Windows ]] Fix linking of engine lcb modules in Windows release mode

This patch changes the export type of the loader classes in
embedded LCB modules (output C mode) to be non-static. This appears
to fix a problem with them not being linked into a release mode
executable on Windows.

runrevmark committed:
Merge pull request #11 from runrevmark/fix-win-release

[[ Windows ]] Fix linking of engine lcb modules in Windows release mode
